File: android/src/main/kotlin/io/agora/agorartm/RTMClient.kt
package io.agora.agorartm

import android.content.Context
import android.os.Handler
import io.agora.rtm.RtmChannel
import io.agora.rtm.RtmClient
import io.agora.rtm.RtmClientListener
import io.agora.rtm.RtmMessage
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

class RTMClient(
    context: Context,
    appId: String?,
    clientIndex: Long,
    messenger: BinaryMessenger,
    private val handler: Handler
) : RtmClientListener, EventChannel.StreamHandler {
    private val eventChannel: EventChannel =
        EventChannel(messenger, "io.agora.rtm.client${clientIndex}")
    private var eventSink: EventChannel.EventSink? = null

    val client: RtmClient?
    val call: RTMCallManager
    val channels: HashMap<String?, RtmChannel> = HashMap()

    init {
        this.eventChannel.setStreamHandler(this)

        this.client = RtmClient.createInstance(context, appId, this)
        this.call = RTMCallManager(client, clientIndex, messenger, handler)
    }

    private fun sendEvent(eventName: String, params: HashMap<Any, Any>) {
        handler.post {
            this.eventSink?.success(
                hashMapOf(
                    "event" to eventName,
                    "data" to params,
                )
            )
        }
    }

    override fun onListen(params: Any?, eventSink: EventChannel.EventSink) {
        this.eventSink = eventSink
    }

    override fun onCancel(params: Any?) {
        this.eventSink = null
    }

    override fun onConnectionStateChanged(state: Int, reason: Int) {
        sendEvent(
            "onConnectionStateChanged", hashMapOf(
                "state" to state,
                "reason" to reason,
            )
        )
    }

    override fun onMessageReceived(message: RtmMessage, peerId: String) {
        sendEvent(
            "onMessageReceived", hashMapOf(
                "message" to message.toJson(),
                "peerId" to peerId,
            )
        )
    }

    override fun onTokenExpired() {
        sendEvent("onTokenExpired", hashMapOf())
    }

    override fun onTokenPrivilegeWillExpire() {
        sendEvent("onTokenPrivilegeWillExpire", hashMapOf())
    }

    override fun onPeersOnlineStatusChanged(peersStatus: Map<String, Int>) {
        sendEvent(
            "onPeersOnlineStatusChanged", hashMapOf(
                "peersStatus" to peersStatus
            )
        )
    }
}

File: ios/Classes/RTMClient.swift
//
//  RTMClient.swift
//  agora_rtm
//
//  Created by LXH on 2023/5/31.
//

import Foundation
import Flutter
import AgoraRtmKit

class RTMClient: NSObject, FlutterStreamHandler, AgoraRtmDelegate {
    var eventChannel: FlutterEventChannel?
    var eventSink: FlutterEventSink?
    
    var client: AgoraRtmKit?
    var call: RTMCallManager?
    var channels: [String: AgoraRtmChannel] = [:]
    
    init(_ appId: String?, _ clientIndex: Int, _ messenger: FlutterBinaryMessenger) throws {
        super.init()
        self.eventChannel = FlutterEventChannel(name: "io.agora.rtm.client\(clientIndex)", binaryMessenger: messenger)
        self.eventChannel?.setStreamHandler(self)
        
        guard let client = AgoraRtmKit(appId: appId ?? "", delegate: self) else {
            throw NSError(domain: "", code: AgoraRtmLoginErrorCode.invalidAppId.rawValue)
        }
        self.client = client
        self.call = RTMCallManager(client, clientIndex, messenger)
    }
    
    deinit {
        channels.forEach { client?.destroyChannel(withId: $0.key) }
        channels.removeAll()
        eventChannel = nil
        client = nil
        call = nil
    }
    
    func sendEvent(eventName: String, params: [String: Any]) {
        let event: [String: Any?] = [
            "event": eventName,
            "data": params,
        ]
        
        eventSink?(event)
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    func rtmKit(_ kit: AgoraRtmKit, connectionStateChanged state: AgoraRtmConnectionState, reason: AgoraRtmConnectionChangeReason) {
        sendEvent(eventName: "onConnectionStateChanged", params: ["state": state.rawValue, "reason": reason.rawValue])
    }
    
    func rtmKit(_ kit: AgoraRtmKit, messageReceived message: AgoraRtmMessage, fromPeer peerId: String) {
        sendEvent(eventName: "onMessageReceived", params: ["message": message.toJson(), "peerId": peerId])
    }
    
    func rtmKit(_ kit: AgoraRtmKit, peersOnlineStatusChanged onlineStatus: [AgoraRtmPeerOnlineStatus]) {
        sendEvent(eventName: "onPeersOnlineStatusChanged", params: ["peersStatus": onlineStatus.reduce(into: [String: Int]()) {
            $0[$1.peerId] = $1.state.rawValue
        }])
    }
    
    func rtmKitTokenDidExpire(_ kit: AgoraRtmKit) {
        sendEvent(eventName: "onTokenExpired", params: [:])
    }
    
    func rtmKitTokenPrivilegeWillExpire(_ kit: AgoraRtmKit) {
        sendEvent(eventName: "onTokenPrivilegeWillExpire", params: [:])
    }
}
